arbor/morph/region.cpp

```diff
@@ -0,0 +1,255 @@
+#include <set>
+#include <string>
+#include <vector>
+
+#include <arbor/morph/error.hpp>
+#include <arbor/morph/locset.hpp>
+#include <arbor/morph/primitives.hpp>
+#include <arbor/morph/region.hpp>
+
+#include "morph/em_morphology.hpp"
+#include "util/span.hpp"
+#include "util/strprintf.hpp"
+#include "util/range.hpp"
+#include "util/rangeutil.hpp"
+
+namespace arb {
+namespace reg {
+
+mcable_list merge(const mcable_list& lhs, const mcable_list& rhs) {
+    mcable_list v;
+    v.resize(lhs.size() + rhs.size());
+    std::merge(lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), v.begin());
+    return v;
+}
+
+bool is_disjoint_union(const mcable& a, const mcable& b) {
+    if (a.branch!=b.branch) return true;
+    return a<b? a.dist_pos<b.prox_pos: b.dist_pos<a.prox_pos;
+}
+
+bool is_nonnull_intersection(const mcable& a, const mcable& b) {
+    if (a==b) return true; // handles special case of spherical branch
+    if (a.branch!=b.branch) return false;
+    return a<b? a.dist_pos>b.prox_pos: b.dist_pos>a.prox_pos;
+}
```

Comment:
This will return false if a<b but a.dist_pos==b.prox_pos; but I believe mcables represent closed intervals, and so this ought to be permissible?

More fundamentally, how does (or should) this differ from just testing `!is_disjoint_union(a, b)`?

Reply:
The problem is that I am effectively ignoring regions that contain a single point.
Given two cables: `a=(cable 1 0 0.5)` and `b=(cable 1 0.5 1)`, the current implementation gives

`(join a b) -> (cable 1 0 1)` ... nothing controversial there
`(intersection a b) -> {}` when it should be `(cable 1 0.5 0.5)` if we treat cable segments as closed intervals.

Using `!is_disjoint(a,b)` makes unit tests fail where we get these interesctions that define a location.

If we allow zero-length regions like this, we would still be inconsistent:
Take `(join (cable p 0 1) (cable c 0 1))` where branch `p` is the parent of child branch `c`: this would still be empty, because the code that performs the intersection performs intersections on cables on the same branch, ignoring parent-child relationships.

I  propose either ignoring zero-length regions, or using half open intervals to keep the current results. Outside of formal completeness, do we have a practical use for compact finite regions?

Reply:
We can define a locset as a multiset on the cover, where we define `m_A(x)` is the multiplicity of location `x` in locset `A`.

The rationale behind using a multiset is:
- If a locset is used to describe the locations of synapses, it has to permit multiple synapses at the same location.
- To describe CV discretisations uniquely with only a set of locations.

How do we define the operations of union (`join`) and intersection (`intersection`)?

To get the "expected" property of intersection distributing over union, and vice versa, i.e.
`A ∪ B) ∩ C = (A ∩ C) ∪ (B ∩ C)` and `A ∩ B) ∪ C = (A ∪ C) ∩ (B ∪ C)`, the following definitions work:
- `C =  A ∪ B`: `m_C(x) = max(m_A(x), m_B(x))`
- `C =  A ∩ B`: `m_C(x) = min(m_A(x), m_B(x))`

However, when a user "combines" two sets of locations to define where to place synapses, they the union above would not provide the behaviour that they expect.
We could implement a `sum` operator:
- `C = A + B`: `m_C(x) = m_A(x) + m_B(x)`

With this, the following "painting" would work logically for a user `cell.paint("expsyn", "sum(A, B)")`.

Reply:
Yup, definitely. For multisets, + and ∪ do have to be distinct. We should support both of these primitive operations, I believe.

Reply:
Yup, definitely. For multisets, + and ∪ do have to be distinct. We should support both of these primitive operations, I believe.

Reply:
Done. I removed `join` and `intersect` from the public API for locset, because they are not used anywhere currently (and I am still trying to wrap my head around where and when we might want to use them).
I have kept their implementation in the back end, and wrote comprehensive tests for them.